api/auth/session.js

```diff
@@ -1,29 +1,121 @@
+  const cookies = new Cookies(req, null);
+  try {
+    logger.silly(req, 'verifying JWT auth token');
+    const valid = jwt.verify(
+      cookies.get(COOKIE_NAME),
+      process.env.SESSION_SECRET,
+      {
+        // Explicitly set the algorithm we expect; otherwise, an attacker could
+        // rewrite the token and set the algorithm to 'none', bypassing our
+        // signature checks
+        algorithms: ['HS256']
+      }
+    );
+
+    // Make sure we issued the token.
+    if (valid.iss !== TOKEN_ISSUER) {
+      throw new Error(`jwt issuer [${valid.iss}] is invalid`);
     }
-  });
+
+    // Return the session contents
+    return valid.payload;
+  } catch (err) {
+    // If the token is invalid, return an empty object rather than failing.
+    // An invalid token does not necessarily mean the request needs to be
+    // terminated (e.g., the token may have expired, so the user needs to
+    // be redirected to a login page).
+    logger.error(req, `invalid auth token: ${err.message}`);
+    return {};
+  }
+};
+
+/**
+ * Returns an Express middleware for handling sessions
+ */
+module.exports = ({ Cookies = defaultCookies } = {}) => {
+  const session = { content: {} };
+
+  /**
+   * Express middleware for managing client-side sessions via JWT
+   * @param {Object} req - Express request object
+   * @param {Object} res - Express response object
+   * @param {Function} next - next() callback
+   */
+  const middleware = (req, res, next) => {
+    const cookies = new Cookies(req, res);
+
+    session.content = loadSession(req);
+
+    // Add a session object to the request and configure it to modify
+    // session.content rather than the actual session object, so we
+    // can be sure the session variable doesn't get mangled
+    Object.defineProperty(req, 'session', {
+      get: () => session.content,
+      set: v => {
+        session.content = v;
+      }
+    });
+
+    // Hijack the response method that writes headers so we can write our
+    // cookie beforehand.
+    const writeHead = res.writeHead;
+    res.writeHead = (...args) => {
+      // Only write a cookie if the session content has stuff in it.
+      if (Object.keys(session.content).length) {
+        logger.silly('writing session cookie to headers');
+
+        // Cookie value is a JWT whose body is an issuer tag and the contents
+        // of session.content as the token payload.  The JWT is signed with
+        // the SESSION_SECRET env var, explicitly sets the algorithm, and sets
+        // an expiration date.
+        cookies.set(
+          COOKIE_NAME,
+          jwt.sign(
+            { iss: 'CMS eAPD API', payload: session.content },
+            process.env.SESSION_SECRET,
+            {
+              algorithm: 'HS256',
+              expiresIn: `${process.env.SESSION_LIFETIME_MINUTES}m`
+            }
+          ),
+          { httpOnly: true }
+        );
+      } else {
+        // Else, write a cookie with an immediate expiration
```

Comment:
what does writing a cookie with an immediate expiration mitigate?

Reply:
It just gets rid of it.  There's no mechanism in the HTTP spec for telling a browser to delete a cookie.  The closest thing is having it expire in the past or right now.